Updates to state migration wiki page (#72)
oppia/oppia-web-developer-docs: Writing-state-migrations.md

## Steps to follow while writing state migrations
1. Make the necessary changes to the State class (or its descendant classes) to reflect the post-migration state structure.

> 2. Increment the CURRENT_STATE_SCHEMA_VERSION in the feconf.py file.
> 3. Increment the CURRENT_EXP_SCHEMA_VERSION in the exp_domain.py file and similar changes in the question_domain.py file.

Reviewer: ```suggestion
2. Increment the `CURRENT_STATE_SCHEMA_VERSION` in the _feconf.py_ file.
3. Increment the `CURRENT_EXP_SCHEMA_VERSION` in the _exp_domain.py_ file and similar changes in the _question_domain.py_ file.
```

Author: Updated

> 4. Start with writing _convert_states_v(old_state_version)_dict_to_v(old_state_version + 1)_dict method in exp_domain.py files under Exploration class and in question_domain.py under Question class. In exp_domain.py, update the ```Exploration._migrate_to_latest_yaml_version``` method to use your conversion function (not required for question_domain.py, this is done automatically).

Reviewer: ```suggestion
4. Start with writing `_convert_states_v{old_state_version}_dict_to_v{old_state_version + 1}_dict` method in _exp_domain.py_ files under `Exploration` class and in _question_domain.py_ under `Question` class. In _exp_domain.py_, update the `Exploration._migrate_to_latest_yaml_version` method to use your conversion function (not required for _question_domain.py_, this is done automatically).
```

Author: Updated

> 5. Write a conversion function _convert_states_v(old_state_version)_dict_to_v(old_state_version + 1)_dict in draft_upgrade_services.py that makes appropriate upgrades to data that resides in ExplorationChange lists. Here you have to check if the changes from your conversion function can be made to the draft_change_list available. If the changes can be made make sure to update the drafts otherwise raise `InvalidDraftConversionException` which will drop the changes in the drafts. If no changes needs to be done to the draft you can simply return the draft_change_list. You will be able to see examples for this in the previously mentioned functions.

Reviewer: ```suggestion
5. Write a conversion function `_convert_states_v{old_state_version}_dict_to_v{old_state_version + 1}_dict` in _draft_upgrade_services.py_ that makes appropriate upgrades to data that resides in `ExplorationChange` lists. Here you have to check if the changes that you do in your conversion function affect the `draft_change_list`, if yes, then make sure to update the drafts or raise `InvalidDraftConversionException` which will drop the changes in the drafts (this needs to be done in cases where changes to drafts would be too complicated). If no changes need to be done to drafts you can simply return the `draft_change_list`.
```

I've removed "You will be able to see examples for this in the previously mentioned functions." because I'm not sure what it means. Can you try to rephrase it?

Author: Updated, I think I do not have to add that line. Just wanted to let the user know that "they can check out previous conversion functions for reference" but they will be able to take reference from the example PRs.

## Steps to follow while writing state migrations
1. Make the necessary changes to the State class (or its descendant classes) to reflect the post-migration state structure.
2. Increment the `CURRENT_STATE_SCHEMA_VERSION` in the _feconf.py_ file.

> 3. Increment the `CURRENT_EXP_SCHEMA_VERSION` in the _exp_domain.py_ file and similar changes in the _question_domain.py_ file.

Reviewer: Hmm, what changes exactly are we doing in the _question_domain.py_?

Author: I have updated this, we do not have to update the question_domain.py for schema version.

> 4. Start with writing `_convert_states_v{old_state_version}_dict_to_v{old_state_version + 1}_dict` method in _exp_domain.py_ files under `Exploration` class and in _question_domain.py_ under `Question` class. In _exp_domain.py_, update the `Exploration._migrate_to_latest_yaml_version` method to use your conversion function (not required for _question_domain.py_, this is done automatically).

Reviewer: ```suggestion
4. Start with writing `_convert_states_v{old_state_version}_dict_to_v{old_state_version + 1}_dict` method in _exp_domain.py_ under `Exploration` class and in _question_domain.py_ under `Question` class. In _exp_domain.py_, update the `Exploration._migrate_to_latest_yaml_version` method to use your conversion function.
```

Author: Done

> 5. Write a conversion function `_convert_states_v{old_state_version}_dict_to_v{old_state_version + 1}_dict` in _draft_upgrade_services.py_ that makes appropriate upgrades to data that resides in `ExplorationChange` lists. Here you have to check if the changes that you do in your conversion function affect the `draft_change_list`, if yes, then make sure to update the drafts or raise `InvalidDraftConversionException` which will drop the changes in the drafts (this needs to be done in cases where changes to drafts would be too complicated). If no changes need to be done to drafts you can simply return the `draft_change_list`.

Reviewer: "that you do in your conversion function" → please explain which conversion function exactly

Maybe split this text into multiple subpoints for better readability

Author: Done

> 6. You would require to make some changes to the existing test files according to the changes you are doing in your conversion function. You have to edit the test functions present for the previous conversion functions in the `exp_domain_test.py` file to update the latest schema version. Also you have to update the schema of core/tests/data/oppia-ThetitleforZIPdownloadhandlertest!-v2-gold.zip file.

Reviewer: ```suggestion
6. Edit the tests for the previous conversion functions in the `exp_domain_test.py` to update to the latest schema version. Also, update the schema of core/tests/data/oppia-ThetitleforZIPdownloadhandlertest!-v2-gold.zip file.
```

Maybe explain how to update the zip.

Reviewer: @lkbhitesh07 Maybe you missed this?

Author: Missed to comment here, I mentioned to unzip the file and then edit it.

> 7. Create a PR. If the tests fail, try resolving the test issues.

Reviewer: ```suggestion
7. Create a PR. If the tests fail resolve them.
```

Author: Done

> 8. Once your PR is finalized, file a one request for the `AuditExplorationMigrationJob` and `MigrateExplorationJob` using this [form](https://docs.google.com/forms/d/e/1FAIpQLSfvYWscAn18ok06An1oQ54h1VmBHfCX8uuuV01kIvY9WX0-Ug/viewform). The job tests a migration by running your conversion function on the dicts of existing exploration models and validating that the migration will be successful. Make sure to mention to only run `MigrateExplorationJob` when the `AuditExplorationMigrationJob` is successful. The audit job does not commit changes to the datastore. After successful testing you can get you PR merged.

Reviewer: ```suggestion
8. Once your PR is finalized, file a one request for the `AuditExplorationMigrationJob` and `MigrateExplorationJob` using this [form](https://docs.google.com/forms/d/e/1FAIpQLSfvYWscAn18ok06An1oQ54h1VmBHfCX8uuuV01kIvY9WX0-Ug/viewform). The job tests a migration by running your conversion function on the dicts of existing exploration models and validating that the migration will be successful. Make sure to mention to only run `MigrateExplorationJob` when the `AuditExplorationMigrationJob` is successful. After successful testing you can get you PR merged.
```

Maybe instead of linking the form directly link the wiki page for testing the jobs.

Author: Do you refer to the Apache beam wiki here?

Reviewer: This one https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production

Author: Updated

2. Increment the `CURRENT_STATE_SCHEMA_VERSION` in the _feconf.py_ file.
3. Increment the `CURRENT_EXP_SCHEMA_VERSION` in the _exp_domain.py_ file.
4. Start with writing `_convert_states_v{old_state_version}_dict_to_v{old_state_version + 1}_dict` method in _exp_domain.py_ under `Exploration` class and in _question_domain.py_ under `Question` class. In _exp_domain.py_, update the `Exploration._migrate_to_latest_yaml_version` method to use your conversion function.
5. Write a conversion function in _draft_upgrade_services.py_
    - Write `_convert_states_v{old_state_version}_dict_to_v{old_state_version + 1}_dict` function that makes appropriate upgrades to data that resides in `ExplorationChange` lists.
    - Here you have to check if the changes that you do in your exploration conversion function affect the `draft_change_list`, if yes, then make sure to update the drafts or raise `InvalidDraftConversionException` which will drop the changes in the drafts (this needs to be done in cases where changes to drafts would be too complicated).
    - If no changes need to be done to drafts you can simply return the `draft_change_list`.

> 6. Edit the tests for the previous conversion functions in the `exp_domain_test.py` to update to the latest schema version. Also, update the schema of core/tests/data/oppia-ThetitleforZIPdownloadhandlertest!-v2-gold.zip file. To update the zip file you would require to unzip it and then make the changes.

Reviewer: ```suggestion
6. Edit the tests for the previous conversion functions in the _exp_domain_test.py_ to update to the latest schema version. Also, update the schema of _core/tests/data/oppia-ThetitleforZIPdownloadhandlertest!-v2-gold.zip_ file. To update the zip file you would require to unzip it and then make the changes.
```

Author: Done

7. Create a PR. If the tests fail resolve them.
8. Once your PR is finalized, file a one request for the `AuditExplorationMigrationJob` and `MigrateExplorationJob` by following the [wiki page](https://github.com/oppia/oppia/wiki/Testing-jobs-and-other-features-on-production). The job tests a migration by running your conversion function on the dicts of existing exploration models and validating that the migration will be successful. Make sure to mention to only run `MigrateExplorationJob` when the `AuditExplorationMigrationJob` is successful. After successful testing you can get you PR merged.

> **Note:** These steps are for the migration where one is changing the schema of all existing states, depending on the changes your migration is going to make the steps will be less as you’ll have to change very fewer test files.
> 

Reviewer 2: Can you convert this to:
```suggestion
> **Note** 
These steps are for the migration where one is changing the schema of all existing states, depending on the changes your migration is going to make the steps will be less as you’ll have to change very fewer test files.
```

Author: Done

If you find new test files where changes needed to be required, try updating the list.

**Links to relevant PRs:**
 - Deprecate math interactions' unsupported rules and update cust arg name: [#15271](https://github.com/oppia/oppia/pull/15271)
 - Add state migration for new customization_arg catchMisspellings: [#16077](https://github.com/oppia/oppia/pull/16077). Please note that in the PR changes are done in several files and those are done according to the requirements.

## Important note:

> - Make sure to add a thorough test for the migration function covering each possible case. Check the PR [#11466](https://github.com/oppia/oppia/pull/11466/files) for reference. Please note that while you are writing tests in the `exp_domain_test.py` file you should use `swap` function to swap the CURRENT_STATE_SCHEMA_VERSION to the schema version you are writing, This way in case you have several tests the next person who will write the schema tests do not have to change `states_schema_version` of every test present. We currently do not have any specific reference to this change, the above PR does not include this as we recently decided to go ahead with this decision. You can always take reference of `swap` function in the codebase to see the usage.

Reviewer: ```suggestion
- Make sure to add a thorough test for the migration function covering each possible case. Check the PR [#11466](https://github.com/oppia/oppia/pull/11466/files) for reference. Please note that while you are writing tests in the _exp_domain_test.py_ file you should use `swap` function to swap the CURRENT_STATE_SCHEMA_VERSION to the schema version you are writing, This way in case you have several tests the next person who will write the schema tests do not have to change `states_schema_version` of every test present. We currently do not have any specific reference to this change, the above PR does not include this as we recently decided to go ahead with this decision. You can always take reference of `swap` function in the codebase to see the usage.
```

Author: Done

- Create a new exploration, make some changes and save them.
- Checkout the feature branch which contains state migration.
- Go to 0.0.0.0:8000 and flush existing Memcache from the Memcache tab.
- Go to the admin page and assign yourself the role of "release coordinator".
- Go to the Misc tab of /release-coordinator page and flush the cache.

> - Run `AuditExplorationMigrationJob` and wait for the job to get completed. This job will not make any changes to the exploration as this is simply an audit job. We run this job before the actual migration job so that in case anything fails we do not make changes to the datastore.

Reviewer: Maybe explain that this should be successful without any errors.

Author: Done

     - core/templates/services/state-interaction-stats.service.spec.ts
     - core/templates/services/state-top-answers-stats.service.spec.ts

7. Create a PR. If the tests fail, try resolving the test issues.

> 8. Once your PR is finalized, file a one off job request for the `AuditExplorationMigrationJob` and `MigrateExplorationJob` using this [form](https://docs.google.com/forms/d/e/1FAIpQLSfvYWscAn18ok06An1oQ54h1VmBHfCX8uuuV01kIvY9WX0-Ug/viewform). The job tests a migration by running your conversion function on the dicts of existing exploration models and validating that the migration will be successful. Make sure to mention to only run `MigrateExplorationJob` when the `AuditExplorationMigrationJob` is successfull. The audit job does not commit changes to the datastore.

Reviewer: ```suggestion
8. Once your PR is finalized, file a job request for the `AuditExplorationMigrationJob` and `MigrateExplorationJob` using this [form](https://docs.google.com/forms/d/e/1FAIpQLSfvYWscAn18ok06An1oQ54h1VmBHfCX8uuuV01kIvY9WX0-Ug/viewform). The job tests a migration by running your conversion function on the dicts of existing exploration models and validating that the migration will be successful. Make sure to mention to only run `MigrateExplorationJob` when the `AuditExplorationMigrationJob` is successfull. The audit job does not commit changes to the datastore.
```

Author: Updated

9. Fix any issues or errors from the audit job above.
10. Get your migration PR merged.

> 11. Once your PR is merged, please submit a request using this [form](https://docs.google.com/forms/d/e/1FAIpQLSfvYWscAn18ok06An1oQ54h1VmBHfCX8uuuV01kIvY9WX0-Ug/viewform) to run this migration in production. Before submitting this request, please ensure that the migration has already been tested manually on your local machine, passed code review, and been merged into develop.

Reviewer: This is not needed? The request in step 8 should handle this.

Author: Updated

**Note:** These steps are for the migration where one is changing the schema of all existing states, depending on the changes your migration is going to make the steps will be less as you’ll have to change very fewer test files.

If you find new test files where changes needed to be required, try updating the list.

**Links to relevant PRs:**
 - Deprecate math interactions' unsupported rules and update cust arg name: [#15271](https://github.com/oppia/oppia/pull/15271)
 - Add state migration for new customization_arg catchMisspellings: [#16077](https://github.com/oppia/oppia/pull/16077). Please note that in the PR changes are done in several files and those are done according to the requirements.

## Important note:

> - Make sure to add a thorough test for the migration function covering each possible case. Check the PR [#11466](https://github.com/oppia/oppia/pull/11466/files) for reference. Please note that while you are writing tests in the `exp_domain_test.py` file you should use `swap` function to swap the CURRENT_STATE_SCHEMA_VERSION to the schema version you are writing, This way in case you have several tests the next person who will write the schema tests do not have to change `states_schema_version` of every test present. We currently do not have any specefic reference to this change, the above PR does not include this as we recently decided to go ahead with this decission. You can always take reference of `swap` function in the codebase to see the usage.

Reviewer: ```suggestion
- Make sure to add a thorough test for the migration function covering each possible case. Check the PR [#11466](https://github.com/oppia/oppia/pull/11466/files) for reference. Please note that while you are writing tests in the `exp_domain_test.py` file you should use `swap` function to swap the `CURRENT_STATE_SCHEMA_VERSION` to the schema version you are writing, This way in case you have several tests the next person who will write the schema tests do not have to change `states_schema_version` of every test present. We currently do not have any specific reference to this change, the above PR does not include this as we recently decided to go ahead with this decision. You can always take reference of `swap` function in the codebase to see the usage.
```

Author: Updated

>      - core/templates/services/state-top-answers-stats.service.spec.ts

Reviewer: Should the list of files be removed?

Author: No, These files may need change according to the requirements. I think this list should be there.

Reviewer: I do not remember you changing those files, also I think a list can be easily outdated, I would rather remove it and explain what needs to be done.

Author: I have removed the list of files and added the description.

If you find new test files where changes needed to be required, try updating the list.

**Links to relevant PRs:**
 - Deprecate math interactions' unsupported rules and update cust arg name: [#15271](https://github.com/oppia/oppia/pull/15271)

>  - Migration related to changing specific interaction schema: [#6177](https://github.com/oppia/oppia/pull/6177). Please note that this PR does not involve changes of ```question_domain``` and ```draft_upgrade_services```.

Reviewer: Remove this, it is quite old, maybe add some newer one (apart from the one on previous line).

Author: I have updated this part.

- Create a new exploration, make some changes and save them.
- Checkout the feature branch which contains state migration.
- Go to 0.0.0.0:8000 and flush existing Memcache from the Memcache tab.

> - Go to admin page and assign yourself the role of `release coordinator`.
> - Go to the Misc tab of release-coordinator page and flush cache.
> - Run AuditExplorationMigrationJob and wait for the job to get completed. This job will not make any changes to the exploration as this is simply an audit job. We run this job before the actual migration job so that in case anything fails we do not make changes to the datastore.
> - Run MigrateExplorationJob and wait for the job to get completed. This will make changes to exploration.

Reviewer: ```suggestion
- Go to the admin page and assign yourself the role of "release coordinator".
- Go to the Misc tab of /release-coordinator page and flush the cache.
- Run `AuditExplorationMigrationJob` and wait for the job to get completed. This job will not make any changes to the exploration as this is simply an audit job. We run this job before the actual migration job so that in case anything fails we do not make changes to the datastore.
- Run `MigrateExplorationJob` and wait for the job to get completed. This will make changes to exploration.
```

Author: Updated

- Check the output of the job and post the screen-shot in your PR.
- Go to the exploration you have created lately, check whether it's working as expected.
- Check demo exploration (note demo exploration ids are 0, 1, 2, etc.)

> - Add a report in your migration PR.

Reviewer: What report?

Author: Updated
